setup_dependent_environment overwritten by module-defined external compiler

## Problem

In Spack, a compiler definition that uses a `module:` token can interfere with or nullify any environment variable defined by `setup_dependent_environment()` or `setup_dependent_package()` of a package upstream in the DAG.

## Question

How does one get around this, with the fewest assumptions and least _a priori_ knowledge?

* Can one defer calling the `setup_dependent_*` hooks until _after_ the compiler is loaded?

  If I'm not mistaken, that's the canonical way of doing things outside of Spack. I can see why Spack would load compilers last, perhaps to largely (though not entirely) guarantee a consistent compiler environment?

* Blacklist "known" variables for `module load` in the specific situation of loading a compiler module? – I think that could only cure specific symptoms.

* Have the user modify the module that defines the compiler? – Ditto for symptoms-only, and I'd say that's impractical for many users.

  On my own cluster, that's what I had done, if unrelated to Spack: I create modulefiles for `intel-parallel-studio` by a script from `psxevars.sh` but _excised_ any MPI variables and paths and spirits those away into a separate `impi` package. This avoided a hit from the main example that follows:

## Main Example: I_MPI_ROOT

The present issue crystallized from the root cause for the failure in issue #8410, which concerns `$I_MPI_ROOT`, defined by both `intel-parallel-studio` (any edition) and `intel-mpi`. 

Consider the scenario where a user runs:
```
spack install clientpkg %intel ^intel-mpi
```
and where the user has in `compilers.yaml`:
``` yaml
- compiler:
    …
    spec:       intel@18.0.3
    # Module provided externally or perhaps even internally by Spack:
    modules:    [intel/18]
    paths:
        …
```

Then, there are two failure modes:

### Failure mode 1: MPI Runtime vs. MPI Development version

When `intel-parallel-studio` is `@composer.*` or `@professional.*`, it includes an MPI runtime, and therefore sets `I_MPI_ROOT`. This value supersedes any specific value or the internal default if unset for any prerequisite `^intel-mpi` version.

When compiling a client package, `mpicc` etc. from `intel-mpi` will error out due to missing static libs only found under `$I_MPI_ROOT` of `intel-mpi`, but not that of `intel-parallel-studio`. Sadly, that information is buried in `config.log`. The user sees the rather more dramatic message:
```
configure: error: C compiler cannot create executables
```

### Failure mode 2: Silent version overshadowing
When `intel-parallel-studio` is `@cluster.*`, a version mismatch occurs for MPI components: In the eyes of the user and Spack, the client package is compiled with `intel-mpi@foo`, whereas `mpicc` (which is from `intel-mpi@foo`) will take its libs etc. from `intel-parallel-studio`.

Disconcertingly, this will occur _silently_.

Now, a user would not typically need or use `^intel-mpi` when one has `%intel-parallel-studio@cluster.*`, but Spack will let you do it. It may not be clear to a new user that it's typically not needed, and it's bad for advanced users who specifically choose a different MPI version.

## Other affected packages: MKL, …

The same is true for the other `intel-foo` library-style packages, of which `intel-mkl` is most critical.

The MKL _is_ provided by all studio versions, so failure mode 1. does not occur, but failure mode 2. might, especially when the user has no access to newer compilers but does for the MKL, which has been available under a "simplified" (no-cost) license since 2017.

## Variable tracing
Here's a more detailed tracing of the MPI variables for the case discussed in  #8410 :
```
spack --debug -v -v install -v -v hdf5@1.10.2+mpi ^intel-mpi
```

Gives (formatted for readability):
```
==> Installing hdf5

==> setup_package.setup_dependent_package:
    spec 'intel-mpi@2018. ...' received .mpi* properties:
    {
	'MPICC':   '.../intel-mpi/2018.3.222-intel-18.0.3/.../intel64/bin/mpiicc',
	'MPICXX':  '.../intel-mpi/2018.3.222-intel-18.0.3/.../intel64/bin/mpiicpc',
	'MPIF77':  '.../intel-mpi/2018.3.222-intel-18.0.3/.../intel64/bin/mpiifort',
	'MPIF90':  '.../intel-mpi/2018.3.222-intel-18.0.3/.../intel64/bin/mpiifort'

	'MPIFC':   '.../intel-mpi/2018.3.222-intel-18.0.3/.../intel64/bin/mpiifort',
    }
...
==> setup_dependent_environment._setup_dependent_env_callback.mpi_setup_dependent_environment:
    adding to spack_env:
    {
	'MPICC':      '.../intel-mpi/2018.3.222-intel-18.0.3/.../intel64/bin/mpiicc',
	'MPICXX':     '.../intel-mpi/2018.3.222-intel-18.0.3/.../intel64/bin/mpiicpc',
	'MPIF77':     '.../intel-mpi/2018.3.222-intel-18.0.3/.../intel64/bin/mpiifort',
	'MPIF90':     '.../intel-mpi/2018.3.222-intel-18.0.3/.../intel64/bin/mpiifort',

	'I_MPI_CC':   '.../lib/spack/env/intel/icc',
	'I_MPI_CXX':  '.../lib/spack/env/intel/icpc',
	'I_MPI_F77':  '.../lib/spack/env/intel/ifort'
	'I_MPI_F90':  '.../lib/spack/env/intel/ifort',
	'I_MPI_FC':   '.../lib/spack/env/intel/ifort',
	'I_MPI_ROOT': '.../.../intel-mpi/2018.3.222-intel-18.0.3/Hipwooo/compilers_and_libraries_2018.3.222/linux/mpi',
    }


==> '/usr/bin/modulecmd' 'python' 'load' 'intel-parallel-studio/composer.2018.3/intel-18.0.3/Hnd7rbb'
...
```

.. and fails with:
```
  >> 33    configure: error: in `/tmp/stern/spack-stage/spack-stage-MQP6Az/hdf5-1.10.2':
  >> 34    configure: error: C compiler cannot create executables
```


```
$ module show intel-parallel-studio/composer.2018.3/intel-18.0.3/Hnd7rbb | sed 's,[ \t:]\+/,\n\t/,g'
```
(formatted for readability):
* MPI:
  ```
  prepend-path	LD_LIBRARY_PATH
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mpi/intel64/lib
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mpi/mic/lib
  prepend-path	CLASSPATH
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mpi/intel64/lib/mpi.jar
  setenv		I_MPI_ROOT
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mpi
  prepend-path	MANPATH
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mpi/man
  prepend-path	PATH
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mpi/intel64/bin
  ```
* MKL:
  ```
  prepend-path	CPATH
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mkl/include
  prepend-path	LD_LIBRARY_PATH
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mkl/lib/intel64_lin
  prepend-path	LIBRARY_PATH
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mkl/lib/intel64_lin
  setenv		MKLROOT
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mkl
  prepend-path	NLSPATH
	.../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mkl/lib/intel64_lin/locale/%l_%t/%N
  prepend-path	PKG_CONFIG_PATH
	  .../intel-parallel-studio/composer.2018.3-intel-18.0.3/.../mkl/bin/pkgconfig
  ```
* likewise for DAAL, IPP, etc.


#### Location where `I_MPI_ROOT` is used
This occurs right on top of `mpiicc `:
```
$ grep -C1 I_MPI  /home/SHARE/soft/spack/spack-stern/opt/spack/linux-centos6-x86_64/\
intel-mpi/2018.3.222-intel-18.0.3/Hipwooo/compilers_and_libraries_2018.3.222/linux/mpi/intel64/bin/mpiicc 
```
namely,
```
prefix=.../intel-mpi/2018.3.222-intel-18.0.3/Hipwooo/compilers_and_libraries_2018.3.222/linux/mpi
# The environment variable I_MPI_ROOT may be used to override installation folder path
if [ -n "$I_MPI_ROOT" ] ; then
    prefix=$I_MPI_ROOT;
fi
```
The lib missing  in  #8410 is specifically mentioned as well; it is available as static lib only, and is linked explicitly even for shared linkage:
```
...
    I_MPI_OTHERLIBS=" $libdir/libmpigi.a"
...
    I_MPI_OTHERLIBS=" -lmpigi"
...
```

--- a/lib/spack/spack/build_environment.py
+++ b/lib/spack/spack/build_environment.py
@@ -669,6 +669,8 @@ def setup_package(pkg, dirty):
     # of Spack, can change environment variables that are relevant to the
     # build of packages. To avoid a polluted environment, preserve the
     # value of a few, selected, environment variables
+    # With the current ordering of environment modifications, this is strictly
+    # unnecessary. Modules affecting these variables will be overwritten anyway
     with preserve_environment('CC', 'CXX', 'FC', 'F77'):
         # All module loads that otherwise would belong in previous
         # functions have to occur after the spack_env object has its

--- a/lib/spack/spack/build_environment.py
+++ b/lib/spack/spack/build_environment.py
@@ -160,7 +160,7 @@ def clean_environment():
     env.unset('CPATH')
     env.unset('LD_RUN_PATH')
     env.unset('DYLD_LIBRARY_PATH')
-    
+
     # Remove any macports installs from the PATH.  The macports ld can
     # cause conflicts with the built-in linker on el capitan.  Solves
     # assembler issues, e.g.:
@@ -171,6 +171,7 @@ def clean_environment():
             env.remove_path('PATH', p)
 
     env.apply_modifications()
+
 
 def set_compiler_environment_variables(pkg, env):
     assert pkg.spec.concrete

--- a/lib/spack/spack/build_environment.py
+++ b/lib/spack/spack/build_environment.py
@@ -146,6 +146,32 @@ class MakeExecutable(Executable):
         return super(MakeExecutable, self).__call__(*args, **kwargs)
 
 
+def clean_environment():
+    # Stuff in here sanitizes the build environment to eliminate
+    # anything the user has set that may interfere. We apply it immediately
+    # unlike the other functions so it doesn't overwrite what the modules load.
+    env = EnvironmentModifications()
+
+    # Remove these vars from the environment during build because they
+    # can affect how some packages find libraries.  We want to make
+    # sure that builds never pull in unintended external dependencies.
+    env.unset('LD_LIBRARY_PATH')
+    env.unset('LIBRARY_PATH')
+    env.unset('CPATH')
+    env.unset('LD_RUN_PATH')
+    env.unset('DYLD_LIBRARY_PATH')
+    
+    # Remove any macports installs from the PATH.  The macports ld can
+    # cause conflicts with the built-in linker on el capitan.  Solves
+    # assembler issues, e.g.:
+    #    suffix or operands invalid for `movq'"
+    path = get_path('PATH')
+    for p in path:
+        if '/macports/' in p:
+            env.remove_path('PATH', p)
+
+    env.apply_modifications()
+
 def set_compiler_environment_variables(pkg, env):
     assert pkg.spec.concrete
     compiler = pkg.compiler
@@ -277,27 +303,6 @@ def set_build_environment_variables(pkg, env, dirty):
 
     # Install root prefix
     env.set(SPACK_INSTALL, spack.store.root)
-
-    # Stuff in here sanitizes the build environment to eliminate
-    # anything the user has set that may interfere.
-    if not dirty:
-        # Remove these vars from the environment during build because they
-        # can affect how some packages find libraries.  We want to make
-        # sure that builds never pull in unintended external dependencies.
-        env.unset('LD_LIBRARY_PATH')
-        env.unset('LIBRARY_PATH')
-        env.unset('CPATH')
-        env.unset('LD_RUN_PATH')
-        env.unset('DYLD_LIBRARY_PATH')
-
-        # Remove any macports installs from the PATH.  The macports ld can
-        # cause conflicts with the built-in linker on el capitan.  Solves
-        # assembler issues, e.g.:
-        #    suffix or operands invalid for `movq'"
-        path = get_path('PATH')
-        for p in path:
-            if '/macports/' in p:
-                env.remove_path('PATH', p)
 
     # Set environment variables if specified for
     # the given compiler
@@ -629,6 +634,30 @@ def setup_package(pkg, dirty):
     spack_env = EnvironmentModifications()
     run_env = EnvironmentModifications()
 
+    if dirty:
+        clean_environment()
+
+    # Loading modules, in particular if they are meant to be used outside
+    # of Spack, can change environment variables that are relevant to the
+    # build of packages. To avoid a polluted environment, preserve the
+    # value of a few, selected, environment variables
+    with preserve_environment('CC', 'CXX', 'FC', 'F77'):
+        # All module loads that otherwise would belong in previous
+        # functions have to occur after the spack_env object has its
+        # modifications applied. Otherwise the environment modifications
+        # could undo module changes, such as unsetting LD_LIBRARY_PATH
+        # after a module changes it.
+        for mod in pkg.compiler.modules:
+            # Fixes issue https://github.com/spack/spack/issues/3153
+            if os.environ.get("CRAY_CPU_TARGET") == "mic-knl":
+                load_module("cce")
+            load_module(mod)
+
+        if pkg.architecture.target.module_name:
+            load_module(pkg.architecture.target.module_name)
+
+        load_external_modules(pkg)
+
     set_compiler_environment_variables(pkg, spack_env)
     set_build_environment_variables(pkg, spack_env, dirty)
     pkg.architecture.platform.setup_platform_environment(pkg, spack_env)
@@ -659,27 +688,6 @@ def setup_package(pkg, dirty):
     # Make sure nothing's strange about the Spack environment.
     validate(spack_env, tty.warn)
     spack_env.apply_modifications()
-
-    # Loading modules, in particular if they are meant to be used outside
-    # of Spack, can change environment variables that are relevant to the
-    # build of packages. To avoid a polluted environment, preserve the
-    # value of a few, selected, environment variables
-    with preserve_environment('CC', 'CXX', 'FC', 'F77'):
-        # All module loads that otherwise would belong in previous
-        # functions have to occur after the spack_env object has its
-        # modifications applied. Otherwise the environment modifications
-        # could undo module changes, such as unsetting LD_LIBRARY_PATH
-        # after a module changes it.
-        for mod in pkg.compiler.modules:
-            # Fixes issue https://github.com/spack/spack/issues/3153
-            if os.environ.get("CRAY_CPU_TARGET") == "mic-knl":
-                load_module("cce")
-            load_module(mod)
-
-        if pkg.architecture.target.module_name:
-            load_module(pkg.architecture.target.module_name)
-
-        load_external_modules(pkg)
 
 
 def fork(pkg, function, dirty, fake):

--- a/lib/spack/spack/build_environment.py
+++ b/lib/spack/spack/build_environment.py
@@ -635,29 +635,8 @@ def setup_package(pkg, dirty):
     spack_env = EnvironmentModifications()
     run_env = EnvironmentModifications()
 
-    if dirty:
+    if not dirty:
         clean_environment()
-
-    # Loading modules, in particular if they are meant to be used outside
-    # of Spack, can change environment variables that are relevant to the
-    # build of packages. To avoid a polluted environment, preserve the
-    # value of a few, selected, environment variables
-    with preserve_environment('CC', 'CXX', 'FC', 'F77'):
-        # All module loads that otherwise would belong in previous
-        # functions have to occur after the spack_env object has its
-        # modifications applied. Otherwise the environment modifications
-        # could undo module changes, such as unsetting LD_LIBRARY_PATH
-        # after a module changes it.
-        for mod in pkg.compiler.modules:
-            # Fixes issue https://github.com/spack/spack/issues/3153
-            if os.environ.get("CRAY_CPU_TARGET") == "mic-knl":
-                load_module("cce")
-            load_module(mod)
-
-        if pkg.architecture.target.module_name:
-            load_module(pkg.architecture.target.module_name)
-
-        load_external_modules(pkg)
 
     set_compiler_environment_variables(pkg, spack_env)
     set_build_environment_variables(pkg, spack_env, dirty)
@@ -685,6 +664,27 @@ def setup_package(pkg, dirty):
 
     set_module_variables_for_package(pkg, pkg.module)
     pkg.setup_environment(spack_env, run_env)
+
+    # Loading modules, in particular if they are meant to be used outside
+    # of Spack, can change environment variables that are relevant to the
+    # build of packages. To avoid a polluted environment, preserve the
+    # value of a few, selected, environment variables
+    with preserve_environment('CC', 'CXX', 'FC', 'F77'):
+        # All module loads that otherwise would belong in previous
+        # functions have to occur after the spack_env object has its
+        # modifications applied. Otherwise the environment modifications
+        # could undo module changes, such as unsetting LD_LIBRARY_PATH
+        # after a module changes it.
+        for mod in pkg.compiler.modules:
+            # Fixes issue https://github.com/spack/spack/issues/3153
+            if os.environ.get("CRAY_CPU_TARGET") == "mic-knl":
+                load_module("cce")
+            load_module(mod)
+
+        if pkg.architecture.target.module_name:
+            load_module(pkg.architecture.target.module_name)
+
+        load_external_modules(pkg)
 
     # Make sure nothing's strange about the Spack environment.
     validate(spack_env, tty.warn)

--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -1478,7 +1478,7 @@ class Spec(object):
         if self.external:
             d['external'] = {
                 'path': self.external_path,
-                'module': bool(self.external_module)
+                'module': self.external_module
             }
 
         # TODO: restore build dependencies here once we have less picky
@@ -1916,7 +1916,7 @@ class Spec(object):
                     dedupe(p + patches))
 
         for s in self.traverse():
-            if s.external_module:
+            if s.external_module and not s.external_path:
                 compiler = spack.compilers.compiler_for_spec(
                     s.compiler, s.architecture)
                 for mod in compiler.modules:

--- a/lib/spack/spack/test/build_environment.py
+++ b/lib/spack/spack/test/build_environment.py
@@ -233,5 +233,4 @@ def test_spack_paths_before_module_paths(config, mock_packages, monkeypatch):
     spack_path = os.path.join(spack.paths.prefix, 'lib/spack/env')
     paths = os.environ['PATH'].split(':')
 
-    assert paths[0] == spack_path
-    assert module_path in paths
+    assert paths.index(spack_path) < paths.index(module_path)

--- a/lib/spack/spack/test/build_environment.py
+++ b/lib/spack/spack/test/build_environment.py
@@ -131,7 +131,6 @@ def test_cc_not_changed_by_modules(monkeypatch):
 
 @pytest.mark.usefixtures('config', 'mock_packages')
 def test_compiler_config_modifications(monkeypatch):
-
     s = spack.spec.Spec('cmake')
     s.concretize()
     pkg = s.package
@@ -209,3 +208,30 @@ def test_compiler_config_modifications(monkeypatch):
     os.environ.pop('PATH_LIST', None)
     os.environ.pop('EMPTY_PATH_LIST', None)
     os.environ.pop('NEW_PATH_LIST', None)
+
+
+@pytest.mark.regression('9107')
+def test_spack_paths_before_module_paths(config, mock_packages, monkeypatch):
+    s = spack.spec.Spec('cmake')
+    s.concretize()
+    pkg = s.package
+
+    module_path = '/path/to/module'
+
+    def _set_wrong_cc(x):
+        os.environ['PATH'] = module_path + ':' + os.environ['PATH']
+
+    monkeypatch.setattr(
+        spack.build_environment, 'load_module', _set_wrong_cc
+    )
+    monkeypatch.setattr(
+        pkg.compiler, 'modules', ['some_module']
+    )
+
+    spack.build_environment.setup_package(pkg, False)
+
+    spack_path = os.path.join(spack.paths.prefix, 'lib/spack/env')
+    paths = os.environ['PATH'].split(':')
+
+    assert paths[0] == spack_path
+    assert module_path in paths

